### PR TITLE
[fragment,hls,f4m,dash,ism] improve fragment downloading

### PR DESCRIPTION
**Pull request by remitamine**
- resume immediately
- no need to concatenate segments and decrypt them on every resume
- no need to save temp files for segments

and for hls downloader:
- no need to download keys for segments that already downloaded

### CODE REVIEW

**Comment by remitamine** [comment]
this should fix the problem in #9512.

**Comment by weedy** [comment]
Please don't let these ideas die devs

**Comment by yan12125** [comment]
Allowing the ```filename``` field to be a stream object can be misleading... How about "filename_or_stream"?

File: youtube_dl/downloader/common.py [hunk 1]
```diff
@@ -327,21 +327,22 @@ def download(self, filename, info_dict):
...
-        # Check file already present
-        if filename != '-' and (nooverwrites_and_exists or continuedl_and_exists):
-            self.report_file_already_downloaded(filename)
-            self._hook_progress({
-                'filename': filename,
-                'status': 'finished',
-                'total_bytes': os.path.getsize(encodeFilename(filename)),
-            })
-            return True
+        if type(filename).__name__ != 'BytesIO':
```
Comment (yan12125) [inline youtube_dl/downloader/common.py]: Should be better with ```isinstanceof(filename, BytesIO)```

Comment (remitamine) [inline youtube_dl/downloader/common.py]: the `BytesIO` is not used directly in the file, so i decided not to import it.

Comment (yan12125) [inline youtube_dl/downloader/common.py]: Or ```if not hasattr(filename, 'write')``` or ```if isinstance(filename, compat_str)```? Checking for an exact type makes the codes less flexible.

**Comment by remitamine** [comment]
> Allowing the filename field to be a stream object can be misleading... How about "filename_or_stream"?

> Or if not hasattr(filename, 'write') or if isinstance(filename, compat_str)? Checking for an exact type makes the codes less flexible.

done.

**Comment by remitamine** [comment]
is there something still missing in the code?

**Comment by yan12125** [comment]
That looks good to me. I've also tested several scenarios and all looked fine. Feel free to merge or rebase.

**Pull request merged**